File: agent/src/lib/apt.ts
import { access } from "node:fs/promises";
import { promisify } from "node:util";
import { execFile } from "node:child_process";

const execFileAsync = promisify(execFile);
const MAX_OUTPUT_CHARS = 4000;

export type AgentSummary = {
  reachable: boolean;
  upgradableCount: number;
  securityCount: number;
  rebootRequired: boolean;
  checkedAt: string;
  outputPreview: string;
};

export class AgentExecutionError extends Error {
  statusCode: number;

  constructor(message: string, statusCode = 500) {
    super(message);
    this.name = "AgentExecutionError";
    this.statusCode = statusCode;
  }
}

function trimOutput(output: string): string {
  const clean = output.trim();
  return clean.length > MAX_OUTPUT_CHARS ? clean.slice(-MAX_OUTPUT_CHARS) : clean;
}

async function runAptCommand(command: string, args: string[]): Promise<string> {
  const result = await execFileAsync(command, args, {
    env: {
      ...process.env,
      DEBIAN_FRONTEND: "noninteractive",
    },
    maxBuffer: 10 * 1024 * 1024,
  });

  return `${result.stdout}\n${result.stderr}`.trim();
}

async function getUpgradableSummary(): Promise<{
  upgradableCount: number;
  securityCount: number;
  rawList: string;
}> {
  const rawList = await runAptCommand("apt", ["list", "--upgradable"]);
  const lines = rawList
    .split(/\r?\n/)
    .map((line) => line.trim())
    .filter((line) => line.length > 0 && !line.startsWith("Listing..."));

  const securityCount = lines.filter((line) => /\/[^ ]*security[^ ]*\s/i.test(line)).length;

  return {
    upgradableCount: lines.length,
    securityCount,
    rawList,
  };
}

async function detectRebootRequired(): Promise<boolean> {
  try {
    await access("/var/run/reboot-required");
    return true;
  } catch {
    return false;
  }
}

function formatSummary(updateOutput: string, listOutput: string, upgradeOutput?: string): string {
  return trimOutput([updateOutput, upgradeOutput, listOutput].filter(Boolean).join("\n\n"));
}

export async function runRefresh(): Promise<AgentSummary> {
  const updateOutput = await runAptCommand("apt-get", ["update"]);
  const packageSummary = await getUpgradableSummary();
  const rebootRequired = await detectRebootRequired();

  return {
    reachable: true,
    upgradableCount: packageSummary.upgradableCount,
    securityCount: packageSummary.securityCount,
    rebootRequired,
    checkedAt: new Date().toISOString(),
    outputPreview: formatSummary(updateOutput, packageSummary.rawList),
  };
}

export async function runUpgrade(allowUpgrade: boolean): Promise<AgentSummary> {
  if (!allowUpgrade) {
    throw new AgentExecutionError("Upgrade is disabled on this agent", 403);
  }

  const updateOutput = await runAptCommand("apt-get", ["update"]);
  const upgradeOutput = await runAptCommand("apt-get", ["upgrade", "-y"]);
  const packageSummary = await getUpgradableSummary();
  const rebootRequired = await detectRebootRequired();

  return {
    reachable: true,
    upgradableCount: packageSummary.upgradableCount,
    securityCount: packageSummary.securityCount,
    rebootRequired,
    checkedAt: new Date().toISOString(),
    outputPreview: formatSummary(updateOutput, packageSummary.rawList, upgradeOutput),
  };
}

File: agent/internal/platform/apt.go
package platform

import (
	"bytes"
	"errors"
	"fmt"
	"os"
	"os/exec"
	"regexp"
	"strings"
	"time"
)

var ErrUpgradeDisabled = errors.New("upgrade is disabled on this agent")

type Summary struct {
	Reachable       bool      `json:"reachable"`
	UpgradableCount int       `json:"upgradableCount"`
	SecurityCount   int       `json:"securityCount"`
	RebootRequired  bool      `json:"rebootRequired"`
	CheckedAt       time.Time `json:"checkedAt"`
	OutputPreview   string    `json:"outputPreview"`
}

var securityPattern = regexp.MustCompile(`/[^ ]*security[^ ]*\s`)

func RunRefresh() (Summary, error) {
	_, err := runCommand("apt-get", "update")
	if err != nil {
		return Summary{}, err
	}

	listOutput, err := runCommand("apt", "list", "--upgradable")
	if err != nil {
		return Summary{}, err
	}

	upgradableCount, securityCount := parseUpgradableList(listOutput)

	return Summary{
		Reachable:       true,
		UpgradableCount: upgradableCount,
		SecurityCount:   securityCount,
		RebootRequired:  fileExists("/var/run/reboot-required"),
		CheckedAt:       time.Now().UTC(),
		OutputPreview:   buildUpgradablePreview(listOutput),
	}, nil
}

func RunUpgrade(allowUpgrade bool) (Summary, error) {
	if !allowUpgrade {
		return Summary{}, ErrUpgradeDisabled
	}

	_, err := runCommand("apt-get", "update")
	if err != nil {
		return Summary{}, err
	}

	_, err = runCommand("apt-get", "upgrade", "-y")
	if err != nil {
		return Summary{}, err
	}

	listOutput, err := runCommand("apt", "list", "--upgradable")
	if err != nil {
		return Summary{}, err
	}

	upgradableCount, securityCount := parseUpgradableList(listOutput)

	return Summary{
		Reachable:       true,
		UpgradableCount: upgradableCount,
		SecurityCount:   securityCount,
		RebootRequired:  fileExists("/var/run/reboot-required"),
		CheckedAt:       time.Now().UTC(),
		OutputPreview:   buildUpgradablePreview(listOutput),
	}, nil
}

func runCommand(command string, args ...string) (string, error) {
	cmd := exec.Command(command, args...)
	cmd.Env = append(os.Environ(), "DEBIAN_FRONTEND=noninteractive")

	var stdout bytes.Buffer
	var stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr

	if err := cmd.Run(); err != nil {
		return "", fmt.Errorf("%s %s failed: %w\n%s", command, strings.Join(args, " "), err, stderr.String())
	}

	return strings.TrimSpace(stdout.String() + "\n" + stderr.String()), nil
}

func parseUpgradableList(output string) (int, int) {
	lines := extractUpgradableLines(output)
	securityCount := 0

	for _, line := range lines {
		if securityPattern.MatchString(line) {
			securityCount++
		}
	}

	return len(lines), securityCount
}

func extractUpgradableLines(output string) []string {
	lines := strings.Split(output, "\n")
	upgradableLines := make([]string, 0, len(lines))

	for _, line := range lines {
		trimmed := strings.TrimSpace(line)
		if trimmed == "" || strings.HasPrefix(trimmed, "Listing...") || strings.HasPrefix(trimmed, "WARNING:") {
			continue
		}

		if !strings.Contains(trimmed, "[upgradable from:") {
			continue
		}

		upgradableLines = append(upgradableLines, trimmed)
	}

	return upgradableLines
}

func buildUpgradablePreview(output string) string {
	lines := extractUpgradableLines(output)
	if len(lines) == 0 {
		return ""
	}

	return trimPreview(strings.Join(lines, "\n"))
}

func trimPreview(value string) string {
	const maxChars = 4000
	trimmed := strings.TrimSpace(value)
	if len(trimmed) <= maxChars {
		return trimmed
	}
	return trimmed[len(trimmed)-maxChars:]
}

func fileExists(path string) bool {
	_, err := os.Stat(path)
	return err == nil
}
